07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
 // This file is part of the "Nabla Engine".
 // For conditions of distribution and use, see copyright notice in nabla.h
 
-
 // I've moved out a tiny part of this example into a shared header for reuse, please open and read it.
 #include "../common/BasicMultiQueueApplication.hpp"
 #include "../common/MonoAssetManagerAndBuiltinResourceApplication.hpp"
 
+//std::this_thread::sleep_for(std::chrono::seconds(1)); // TODO: remove
+
 using namespace nbl;
 using namespace core;
 using namespace system;
 using namespace asset;
 using namespace video;
 
-
 #include "app_resources/common.hlsl"
 
-
 // This time we let the new base class score and pick queue families, as well as initialize `nbl::video::IUtilities` for us
 class StagingAndMultipleQueuesApp final : public examples::BasicMultiQueueApplication, public examples::MonoAssetManagerAndBuiltinResourceApplication
 {
-		using device_base_t = examples::BasicMultiQueueApplication;
-		using asset_base_t = examples::MonoAssetManagerAndBuiltinResourceApplication;
+	using device_base_t = examples::BasicMultiQueueApplication;
+	using asset_base_t = examples::MonoAssetManagerAndBuiltinResourceApplication;
+
+	static constexpr std::array imagesToLoad = {
+		"../app_resources/test0.png",
+		"../app_resources/test1.png",
+		"../app_resources/test2.png",
+		"../app_resources/test0.png",
+		"../app_resources/test1.png",
+		"../app_resources/test2.png",
+		"../app_resources/test1.png",
+		"../app_resources/test2.png",
+		"../app_resources/test0.png"
+	};
+	static constexpr size_t IMAGE_CNT = imagesToLoad.size();
+
+public:
+	// Yay thanks to multiple inheritance we cannot forward ctors anymore
+	StagingAndMultipleQueuesApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
+
+	// This time we will load images and compute their histograms and output them as CSV
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		// Remember to call the base class initialization!
+		if (!device_base_t::onAppInitialized(std::move(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		constexpr size_t TIMELINE_SEMAPHORE_STARTING_VALUE = 0;
+		m_imagesLoadedSemaphore = m_device->createSemaphore(TIMELINE_SEMAPHORE_STARTING_VALUE);
+		m_imagesProcessedSemaphore = m_device->createSemaphore(TIMELINE_SEMAPHORE_STARTING_VALUE);
+		m_histogramSavedSeamphore = m_device->createSemaphore(TIMELINE_SEMAPHORE_STARTING_VALUE);
+
+		// TODO: create/initialize array of atomic pointers to IGPUImage* and IGPUBuffer* to hold results
+		// no need i think
```

Comment:
your `std::array<core::smart_refctd_ptr<IGPUImage>, IMAGE_CNT>` is fine, but whether the pointers need to be atomic depends on how you sync between `loadImages` and `calculateHistogram`